### PR TITLE
osdocs-339-devexp-222 Document setup for additional trusted CAs

### PR DESCRIPTION
**Pull request by bmcelvee**
https://jira.coreos.com/browse/OSDOCS-339
(https://jira.coreos.com/browse/DEVEXP-222)

### CODE REVIEW

**Comment by bmcelvee** [comment]
@adambkaplan PTAL when you get a chance, thanks!

**Comment by adambkaplan** [comment]
@bmcelvee gist of what to do:

Prerequisite - cluster admin must have access to the registry's public certificates, usually a `hostname/ca.crt` file located in `/etc/docker/certs.d/` directory. See [docker reference](https://docs.docker.com/engine/security/certificates/).

Procedure

1. Create a ConfigMap in the `openshift-config` namespace containing the trusted certificates for the registries which use self-signed certificates. For each certificate authority file, make sure the key in the ConfigMap is the registry's hostname in the hostname[..port] format:
```
$ oc create configmap cluster-cas -n openshift-config \
--from-file=myregistry.corp.com..5000=/etc/docker/certs.d/myregistry.corp.com:5000/ca.crt \
--from-file=otherregistry.com=/etc/docker/certs.d/otherregistry.com/ca.crt
```

2. Update the cluster image figuration:
```
$ oc patch image.config.openshift.io/cluster --patch '{"spec":{"additionalTrustedCA":{"name":"cluster-cas"}}}' --type=merge
```

**Comment by bmcelvee** [comment]
Thanks!

File: modules/configmap-create.adoc [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+You can use the following command to create a `ConfigMap` from
+directories, specific files, or literal values.
+
+.Procedure
+
+* Create a ConfigMap:
+
+----
+$ oc create configmap <configmap_name> [options]
```
Comment (adambkaplan) [inline modules/configmap-create.adoc]: fwiw there's a whole host of upstream docs:

https://kubectl.docs.kubernetes.io/pages/app_management/secrets_and_configmaps.html
https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/#create-a-configmap

**Review by adambkaplan — CHANGES_REQUESTED** [review]
A few additional nits. May be worth referencing some of the upstream docs, since the CLI is inherited from kubectl.

**Comment by bmcelvee** [comment]
@wewang58 PTAL, thanks!

**Comment by bmcelvee** [comment]
@openshift/team-documentation please peer review

**Comment by bmcelvee** [comment]
@openshift/team-documentation please peer review, thanks!

File: builds/setting-up-trusted-ca.adoc [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+//Required modules:
+//Cluster admin creates a ConfigMap (1)
+//Add additional CAs (2)
+//--No longer needs single PEM bundle
+//--Each CA must be associated with a domain. Domain format is hostname[..port]
+//--Domain is the key in the ConfigMap - value is the PEM-encoded certificate
+//Cluster admin (or installer?) creates the ConfigMap (redundant??) (1)
+//--ConfigMap name must be set in the following locations
+//  --build.config.openshift.io/cluster
```
[the remark this answers is not in the document]
Comment (adambkaplan) [reply builds/setting-up-trusted-ca.adoc]: @bmcelvee please remove this line

File: modules/configmap-overview.adoc [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+binaryData:
+  bar: L3Jvb3QvMTAw <2>
+----
+<1> Contains the configuration data.
+<2> Points to a file that contains non-UTF8 data, for example, a binary Java keystore file.
+Enter the file path in Base 64.
+
+[NOTE]
+====
+You can use the `binaryData` field when you create a configmap from a file.
```
Comment (adambkaplan) [inline modules/configmap-overview.adoc]:
```suggestion
You can use the `binaryData` field when you create a configmap from a binary file, such as an image.
```

File: modules/configmap-overview.adoc [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  example.property.file: |-
+    property.1=value-1
+    property.2=value-2
+    property.3=value-3
+binaryData:
+  bar: L3Jvb3QvMTAw <2>
+----
+<1> Contains the configuration data.
+<2> Points to a file that contains non-UTF8 data, for example, a binary Java keystore file.
+Enter the file path in Base 64.
```
Comment (adambkaplan) [inline modules/configmap-overview.adoc]:
```suggestion
Enter the file data in Base 64.
```

**Review by adambkaplan — COMMENTED** [review]
Additional nits

**Comment by bmcelvee** [comment]
@openshift/team-documentation please peer review

@wewang58 please take a look

Thanks!

File: builds/setting-up-trusted-ca.adoc [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+= Setting up additional trusted certifying authorities for builds
+include::modules/common-attributes.adoc[]
+:context: setting-up-trusted-ca
+toc::[]
+
+Use the following sections to set up additional certifying authorities (CA).
+
+In general, a cluster admin creates a ConfigMap and adds additional CAs.
+
+* No longer needs single PEM bundle
```
Comment (ahardin-rh) [inline builds/setting-up-trusted-ca.adoc]: What no longer needs a single PEM bundle? I would make this a complete sentence and add a period, to keep consistent with the rest of the list.

File: builds/setting-up-trusted-ca.adoc [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+* Each CA must be associated with a domain. Domain format is `hostname[..port]`.
+* `domain` is the key in the ConfigMap, `value` is the PEM-encoded certificate.
```
Comment (ahardin-rh) [inline builds/setting-up-trusted-ca.adoc]: I would replace the comma after ConfigMap with a semicolon

File: modules/configmap-adding-ca.adoc [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+* You must have cluster administrator privileges.
+* You must have access to the registry's public certificates, usually a
+`hostname/ca.crt` file located in the `/etc/docker/certs.d/` directory.
+
+.Procedure
+
+. Create a ConfigMap in the `openshift-config` namespace containing the trusted
+certificates for the registries that use self-signed certificates. For each
+CA file, make sure the key in the ConfigMap is the registry's
```
Comment (ahardin-rh) [inline modules/configmap-adding-ca.adoc]: s/make sure/ensure

File: builds/setting-up-trusted-ca.adoc [hunk 8]
```diff
@@ -0,0 +1,26 @@
+[id="setting-up-trusted-ca"]
+= Setting up additional trusted certifying authorities for builds
+include::modules/common-attributes.adoc[]
+:context: setting-up-trusted-ca
+toc::[]
+
+Use the following sections to set up additional certifying authorities (CA).
+
+In general, a cluster admin creates a ConfigMap and adds additional CAs.
```
Comment (ahardin-rh) [inline builds/setting-up-trusted-ca.adoc]: s/cluster admin/cluster administrator

**Comment by ahardin-rh** [comment]
@bmcelvee Just a few minor comments from me :+1:

File: modules/configmap-create.adoc (hunk 1 above)
Comment (bmcelvee) [reply modules/configmap-create.adoc]: Added to additional resources :)

**Comment by vikram-redhat** [comment]
@wewang58 PTAL. 

cc: @xltian - can you assign someone else if if Wen is not available.

File: modules/configmap-create.adoc [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+
+You can use the following command to create a ConfigMap from
+directories, specific files, or literal values.
+
+.Procedure
+
+* Create a ConfigMap:
+
+----
+$ oc create configmap <configmap_name> [options]
```
Comment (wzheng1) [inline modules/configmap-create.adoc]: The command should be like：
$oc create configmap registry-config --from-file=docker-registry-default.apps.0128-49c.qe.rhcloud.com=ca.crt -n **openshift-config**

Comment (wzheng1) [reply modules/configmap-create.adoc]: I think this part is duplicated with above part.

**Comment by wzheng1** [comment]
QE will ack this pr, thanks!

**Comment by wzheng1** [comment]
LGTM for QE.

**Comment by bmcelvee** [comment]
@wzheng1 thank you! :)

**Pull request merged**